techmatters/hrm
# feat: Support “valid until X DATE” for blocking

## Description
This PR adds logic to support "valid until" in the associations between profiles and flags (e.g. a client is blocked 24 hours, or what is the same "until tomorrow").

This PR:
- Adds a migration to alter `ProfilesToProfileFlags` table, adding `validUntil` column.
- "Associate" endpoint (`POST` on `profiles/:profileId/flags/:profileFlagId`) accepts an optional `validUntil` query parameter.
  - If missing, the association is saved as permanent (`validUntil` = `null`).
  - If provided, the endpoint will assert that this is a valid date and pass it along `associateProfileToProfileFlag` service.
  - `associateProfileToProfileFlag` service will assert that the date provided (if any) is a future date.
- Update endpoints where profile flags are retrieved with the profile payload (`getProfileById` & `listProfiles`). The payload sent back to the client slightly changed for this two endpoints: before `profileFlags` was an array of identifiers, now it looks like (FYI @mythilytm).
```
[
  {
    "validUntil": "2024-01-05T20:03:12.853447+00:00",
    "id": 1
  },
  {
    "validUntil": null,
    "id": 2
  }
]
```

### Checklist

### Related Issues
Fixes #....

### Verification steps
Trust the service tests or
- Start local server.
- Save a new contact in order to create a new profile.
- Get the profile id.
- Associate the profile to flag id `1` (by doing a `POST` against `/profiles/:profileId/flags/:profileFlagId?validUntil=<date string>`, where `validUntil=<date string>` is the query param used for the valid until column.
- Confirm that the profile - profile flag association is saved in the DB with the specified valid until date.
